### DISCUSSION TITLE
General Structure ideas and decisions

### DISCUSSION THREAD

**Discussion by MDr164** [fstart-io/fstart#16] [General]
We already sorted out that everything should be describable using DT nodes. Now looking at some specifics here: Should the entry point be chooseable as well or should there be an architecture specific entry that all platforms of that arch can/should use? I'm asking because I got a few platforms here ready to start working on driver nodes for but we'd need something at the entry ready in that case.

**Comment by ArthurHeymans** [comment]
So for most platforms a per arch entry point is possible.
I think with global_asm! + format it should be possible to construct a universal and flexible entry point.

**Reply by MDr164** [reply]
I assume we'll have to get RAM up first for a stack or to we want to do something similar as coreboot does by using the CPU caches?

**Reply by i-c-o-n** [reply]
I guess everything but x86 has SRAM that should be just there. So, assembly only has to set up a little environment, point stack-frame register to a linker symbol, etc. I wouldn't expect platform-specific code there.

On x86 I see three possible ways:
* Do raminit in assembly (not feasible for modern platforms; I could try to do that with a VIA C3 board, but I'd expect that would only be for educational purpose).
* Enable cache as RAM in assembly (for medium-aged and bummer platforms?).
* Host processor starts with DRAM enabled (AMD Zen).

So AMD Zen probably wouldn't need platform-specific assembly (I assume the code path could be shared with a QEMU port?). Other x86 does. If there is interest in early platform diversity, I could probably provide ports for the three cases.

**Reply by ArthurHeymans** [reply]
Right. Cache as ram can be a little platform specific but I think it's possible to still tie that to FDT nodes. With rust global_asm!() and some other buildtime code you could have all the code you need before running rust.

**Reply by heijligen** [reply]
U-Boot / Barebox and the Linux kernel are using micro architectures to have an abstraction between individual boards and a whole architecture family. This might fit in here also quite well. E.g. with the idea of a multi board build.
